src/webcore/executor.rs

```diff
@@ -0,0 +1,194 @@
+        EVENT_LOOP_INNER.with(EventLoopInner::drain)
+    }
+    fn push_task(&self, task: Rc< SpawnedTask >) {
+        EVENT_LOOP_INNER.with(|inner| inner.push_task(task))
+    }
+}
+
+// State relating to the javascript event loop. Only one instance ever exists.
+struct EventLoopInner {
+    // Avoid unnecessary allocation and interop by keeping a local
+    // queue of pending tasks.
+    microtask_queue: RefCell< VecDeque< Rc< SpawnedTask > > >,
+    waker: Reference,
+}
+
+// Not strictly necessary, but may become relevant in the future
+impl Drop for EventLoopInner {
+    #[inline]
+    fn drop( &mut self ) {
+        js! { @(no_return)
+            @{&self.waker}.drop();
+        }
+    }
+}
+
+impl EventLoopInner {
+    // Initializes the event loop. Only called once.
+    fn new() -> Self {
+        EventLoopInner {
+            microtask_queue: RefCell::new(VecDeque::with_capacity(INITIAL_QUEUE_CAPACITY)),
+            waker: js!(
+                var callback = @{|| EventLoop.drain()};
+                var wrapper = function() { callback() };
+
+                // Modern browsers can use `MutationObserver` which allows
+                // us to schedule a micro-task without allocating a promise.
+                // https://dom.spec.whatwg.org/#notify-mutation-observers
+                if ( typeof MutationObserver === "function" ) {
+                    var node = document.createTextNode( "0" );
+                    var state = false;
+
+                    new MutationObserver( wrapper ).observe( node, { characterData: true } );
+
+                    function nextTick() {
+                        state = !state;
+                        node.data = ( state ? "1" : "0" );
+                    }
+                    nextTick.drop = callback.drop;
+
+                    return nextTick;
+
+                // Node.js and other environments
+                } else {
+                    var promise = Promise.resolve( null );
+
+                    function nextTick( value ) {
+                        promise.then( wrapper );
+                    }
+                    nextTick.drop = callback.drop;
+
+                    return nextTick;
+                }
+            ).try_into().unwrap()
+        }
+    }
+    // Pushes a task onto the queue
+    fn push_task(&self, task: Rc< SpawnedTask >) {
+        let mut queue = self.microtask_queue.borrow_mut();
+        queue.push_back(task);
+
+        // If the queue was previously empty, then we need to schedule
+        // the queue to be drained.
+        if queue.len() == 1 {
+            self.wake();
+        }
+    }
+    // Invoke the javascript waker function
+    fn wake(&self) {
+        js! { @(no_return) @{&self.waker}(); }
+    }
+    // Remove and return a task from the front of the queue
+    fn pop_task(&self) -> Option< Rc< SpawnedTask > > {
+        self.microtask_queue.borrow_mut().pop_front()
+    }
+    // Poll the queue until it is empty
+    fn drain(&self) {
```

Comment:
This should clean up the memory for the queue. Otherwise it will waste memory if a large number of Tasks are queued (since it will grow the memory, but it will never deallocate the memory).

I had an idea of only cleaning up the queue memory if a certain amount of time has passed (e.g. 5 seconds).

Reply:
Interesting, but I think that can wait for a second PR. I'd prefer not to tie it to wall-clock time, perhaps "shrinking the queue if the maximum fill ratio from the last N iterations drops below 50%" would be a good metric.

Reply:
In that case why not just add back in my implementation of always cleaning the queue? That's simple to add and we can always refine it later.